### ISSUE TITLE
New rule proposal:  no-keyword

### ISSUE THREAD

**Issue by prettydiff** [eslint/eslint#11645]
## Rule example

```json
"no-keyword": [
    ["with", "var", "console"],
    "error"
]
```

## Purpose

This rule will allow a list of keywords and will warn or error when the parser encounters those keywords in the analyzed code.  This rule is less opinionated and more flexible than existing approaches and replaces the following rules with one single rule:

* no-with
* no-var
* no-console
* no-alert
* no-eval
* no-new
* no-void

I have seen several closed issues about rule proposals for certain keywords that were ultimately deemed too opinionated or too narrow.  A rule like this proposal will allow code authors the flexibility to impose their opinions upon their code, but without subjecting those opinions onto the larger ESLint community.

## What category of rule is this? (place an "X" next to just one item)

[ ] Warns about a potential error (problem)
[x] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

### Rule
```json
"no-keyword":[
    ["switch"],
    "error"
]
```

### Code sample
```js
switch (test) {
    case "apple": something; break;
}
```

## Why should this rule be included in ESLint (instead of a plugin)?

This approach is more flexible, extensible, and less opinionated than the existing approach.

## Are you willing to submit a pull request to implement this rule?

Yes, I am willing to do the work, but I am a ESLint newb.

**Comment by platinumazure** [comment]
Hi @prettydiff, thanks for the issue.

You can use [`no-restricted-syntax`](https://eslint.org/docs/rules/no-restricted-syntax) to achieve pretty much the same thing.

Try these selectors:

* `WithStatement`
* `VariableDeclaration[kind='var']`
* `MemberExpression[object.name='console'][property.name=/^(log|info|warn|error)$/]`
* `Identifier[name='alert']`
* `Identifier [name='eval']`
* `NewExpression`
* `UnaryExpression[operator='void']`

Since this (and much more) is already possible with `no-restricted-syntax`, I don't think the proposed rule is necessary.

**Comment by prettydiff** [comment]
Thanks @platinumazure .  How would I restrict use of `new`, `class`, `this`, and `try`?

**Comment by platinumazure** [comment]
@prettydiff `ClassDeclaration, ClassExpression`, `ThisExpression`, `TryStatement`.

I highly recommend using astexplorer.net to see how any given syntax looks when parsed by espree (ESLint's parser), and you can look at node names as a starting point. Note that some tweaking/filtering is sometimes needed. See the `no-restricted-syntax` documentation as a starting point; it also links to esquery documentation, which is sort of like CSS selectors but for the ESTree spec, and it will show you how to get more expressiveness in your `no-restricted-syntax` configuration.

**Comment by prettydiff** [comment]
Thanks.
